README.md

```diff
@@ -1372,6 +1384,298 @@ module.exports = {
+      canonicalValuesMap[identifier] = classNames;
+
+      allExportsJson[resourcePath] = allExportsJson[resourcePath] || {};
+      allExportsJson[resourcePath][localName] = classNames;
+    }
+  }
+}
+
+function replaceReplacements(classNames) {
+  const adjustedClassNames = classNames.replaceAll(
+    REPLACEMENT_REGEX,
+    (_, resourcePath, localName) => {
+      const identifier = generateIdentifier(resourcePath, localName);
+      if (identifier in canonicalValuesMap) {
+        return canonicalValuesMap[identifier];
+      }
+
+      // recurse through other stand-in that may be imports
+      const canonicalValue = replaceReplacements(replacementsMap[identifier]);
+      canonicalValuesMap[identifier] = canonicalValue;
+      return canonicalValue;
+    }
+  );
+
+  return adjustedClassNames;
+}
+
+module.exports = {
+  module: {
+    rules: [
+      {
+        test: /\.css$/i,
+        loader: "css-loader",
+        options: {
+          modules: {
+            getJSON: ({ resourcePath, imports, exports, replacements }) => {
+              const exportsJson = exports.reduce(
+                (acc, { name, value }) => ({ ...acc, [name]: value }),
+                {}
+              );
+
+              if (replacements.length > 0) {
+                // replacements present --> add stand-in values for absolute paths and local names,
+                // which will be resolved to their canonical values in the plugin below
+                addReplacements(
+                  resourcePath,
+                  imports,
+                  exportsJson,
+                  replacements
+                );
+              } else {
+                // no replacements present --> add to canonicalValuesMap verbatim
+                // since all values here are canonical/don't need resolution
+                for (const [key, value] of Object.entries(exportsJson)) {
+                  const id = `[${resourcePath}][${key}]`;
+
+                  canonicalValuesMap[id] = value;
+                }
+
+                allExportsJson[resourcePath] = exportsJson;
+              }
+            },
+          },
+        },
+      },
+    ],
+  },
+  plugins: [
+    {
+      apply(compiler) {
+        compiler.hooks.done.tap("CssModulesJsonPlugin", () => {
+          for (const [identifier, classNames] of Object.entries(
+            replacementsMap
+          )) {
+            const adjustedClassNames = replaceReplacements(classNames);
+            replacementsMap[identifier] = adjustedClassNames;
+            const [, resourcePath, localName] =
+              identifier.match(IDENTIFIER_REGEX);
+            allExportsJson[resourcePath] = allExportsJson[resourcePath] || {};
+            allExportsJson[resourcePath][localName] = adjustedClassNames;
+          }
+
+          fs.writeFileSync(
+            "./output.css.json",
+            JSON.stringify(allExportsJson, null, 2),
+            "utf8"
+          );
+        });
+      },
+    },
+  ],
+};
```

Comment:
Let's keep only this, because other developers can use another solution and faced with the problem better avoid it for better DX

Reply:
And can you add a test case with your solution and put it in `helpers` directory

Reply:
@alexander-akait Sure, I can definitely add a test case!

Regarding the README changes, which parts are you suggesting we keep versus remove? I'm happy to update whatever, but I'm not sure which part you're referring to.

Reply:
I mean keep only part where we replace `___REPLACEMENT`

Reply:
> The third example has a lot more logic involved than I would have liked, but I'm not sure it'd make sense without both the loader and plugin parts.

Let's keep it, composes is a popular things, so better developer will use it, maybe someone send a PR with some optimizations in future

Reply:
Sounds good to me! Thanks for your understanding 🙏